dali/operators/reader/fits_reader_gpu_op.cu

```diff
@@ -0,0 +1,216 @@
+
+      b &= (1 << nbits) - 1;
+      imax = i + blocksize;
+      if (imax > tile_size[tile]) {
+        imax = tile_size[tile];
+      }
+      if (fs < 0) {
+        for (; i < imax; i++) {
+          uncompressed_data[beg + i] = lastpix;
+        }
+      } else if (fs == fsmax[bytelog2]) {
+        for (; i < imax; i++) {
+          k = bbits - nbits;
+          diff = b << k;
+          for (k -= 8; k >= 0; k -= 8) {
+            b = *compressed_tile++;
+            diff |= b << k;
+          }
+          if (nbits > 0) {
+            b = *compressed_tile++;
+            diff |= b >> (-k);
+            b &= (1 << nbits) - 1;
+          } else {
+            b = 0;
+          }
+
+          if ((diff & 1) == 0) {
+            diff = diff >> 1;
+          } else {
+            diff = ~(diff >> 1);
+          }
+          uncompressed_data[beg + i] = diff + lastpix;
+          lastpix = uncompressed_data[beg + i];
+        }
+      } else {
+        for (; i < imax; i++) {
+          while (b == 0) {
+            nbits += 8;
+            b = *compressed_tile++;
+          }
+          nzero = nbits - nonzero_count[b];
+          nbits -= nzero + 1;
+          b ^= 1 << nbits;
+          nbits -= fs;
+          while (nbits < 0) {
+            b = (b << 8) | (*compressed_tile++);
+            nbits += 8;
+          }
+          diff = (nzero << fs) | (b >> nbits);
+          b &= (1 << nbits) - 1;
+
+          if ((diff & 1) == 0) {
+            diff = diff >> 1;
+          } else {
+            diff = ~(diff >> 1);
+          }
+          uncompressed_data[beg + i] = diff + lastpix;
+          lastpix = uncompressed_data[beg + i];
+        }
+      }
+    }
+
+    if (bscale != 1. || bzero != 0.) {
+      for (int j = 0; j < tile_size[tile]; ++j) {
+        uncompressed_data[beg + j] = static_cast<std::make_signed_t<T>>(
+            static_cast<std::make_signed_t<T>>(uncompressed_data[beg + j]) * bscale + bzero);
+      }
+    }
+  }
+}
+
+void FitsReaderGPU::RunImpl(Workspace &ws) {
+  int num_outputs = ws.NumOutput();
+  int batch_size = GetCurrBatchSize();
+  TensorList<CPUBackend> sample_list_cpu;
+
+  kernels::DynamicScratchpad s({}, ws.stream());
+
+  for (int output_idx = 0; output_idx < num_outputs; output_idx++) {
+    auto &output = ws.Output<GPUBackend>(output_idx);
+    auto &sample_0 = GetSample(0);
+    bool compressed = sample_0.header[output_idx].compressed;
+
+    sample_list_cpu.Reset();
+    sample_list_cpu.set_sample_dim(sample_0.data[output_idx].ndim());
+    sample_list_cpu.set_type(sample_0.data[output_idx].type());
+    sample_list_cpu.set_device_id(sample_0.data[output_idx].device_id());
+    sample_list_cpu.SetSize(batch_size);
+
+    for (int sample_id = 0; sample_id < batch_size; ++sample_id) {
+      auto &sample = GetSample(sample_id);
+      sample_list_cpu.SetSample(sample_id, sample.data[output_idx]);
+    }
+
+
+    if (compressed) {
+      TensorList<GPUBackend> sample_list_gpu;
```

Comment:
```suggestion
      TensorList<GPUBackend> sample_list_gpu;
      sample_list_gpu.set_order(ws.stream());
```
I think with that change we can stick with the implementation using the Copy from one TL to another.
It is possible to use scratchpad, but the copy would still be quite manual, I think we can live with this version for now, unless this poses a performance problem, we can take a look into first coalescing it into one buffer.
Let's stick with this version (+ the setting of the order for the temporary buffer - it will ensure that we release the memory on the correct stream).